text/0000-route-serializers.md

```diff
@@ -0,0 +1,111 @@
+
+# Motivation
+
+As we move towards an increasingly asynchronous world with Engines, we need to separate knowledge about how to _**link** to a route_ and how to _**enter** a route_. Linking to a route should be able to happen _before_ a route object is instantiated, which is the behavior needed to asynchronously load Engines. However, in our current reality, these concerns are coupled and a route object must be instantiated before being able to link to _or_ enter a route.
+
+By separating these concerns, we can preemptively load the information on how to link to a route without also requiring all the knowledge of how to enter that route. This would be beneficial in both the asynchronous and synchronous worlds by allowing us to defer work.
+
+Since the `serialize` method is the only method currently used by the `Route` class to define how to link to a route, the proposal is to extract this method into the space which currently contains the other linking information (e.g., the Router's map).
+
+_**Note:** this separation of concerns will also need to be implemented in router.js for the preemptive loading proposed here to actually work, but we can prepare for that future world by creating a separation of concerns within application code now._
+
+# Detailed Design
+
+Since the current API is a simple function, the new hash option will also be a simple function that mirrors the signature of the original. Here's an example:
+
+```js
+// app/router.js
+export default Router.map(function() {
+  this.route('post', {
+    path: '/post/:id',
+    serialize(model, params) {
+      // serialize the model into the dynamic paths
+    }
+  });
+});
+```
+
+Preserving the current function signature means that refactoring existing code should be simple in most cases. Here's the example currently given in the Ember docs (updated to reflect Ember-CLI):
+
+```js
+// app/routes/post.js
+import Ember from 'ember';
+export default Ember.Route.extend({
+  model(params) {
+    // the server returns `{ id: 12 }`
+    return Ember.$.getJSON('/posts/' + params.post_id);
+  },
+
+  serialize(model) {
+    // this will make the URL `/posts/12`
+    return { post_id: model.id };
+  }
+});
+
+// app/router.js
+export default Router.map(function() {
+  this.route('post', {
+    path: '/post/:id'
+  });
+});
+```
+
+Here is that same code refactored for the proposal:
+
+```js
+// app/routes/post.js
+import Ember from 'ember';
+export default Ember.Route.extend({
+  model(params) {
+    // the server returns `{ id: 12 }`
+    return Ember.$.getJSON('/posts/' + params.post_id);
+  }
+});
+
+// app/router.js
+export default Router.map(function() {
+  this.route('post', {
+    path: '/post/:id',
+    serialize(model) {
+      // this will make the URL `/posts/12`
+      return { post_id: model.id };
+    }
+  });
+});
+```
```

Comment:
The core team discussed the pedagogy and strongly favors separating the definition of the `serialize` function from the main route map. This is analogous to how we separate the declaration of `initialize` functions in initializers. 

@trentmwillis could you please change the example above to something like:

```
// app/router.js

function serializePostRoute(model) {
  // this will make the URL `/posts/12`
  return { post_id: model.id };
}

export default Router.map(function() {
  this.route('post', { path: '/post/:id', serialize: serializePostRoute });
});
```